File: dialogio/pas/Menu2_refresh.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2025 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

// From `MenuLib2' unit

procedure refresh;

begin
  _dbg_enter ({$I %FILE%}, 'refresh');

  If (page = opage) and (idx2 = opos) and NOT MenuLib2_mn_environment.do_refresh then
    begin
      _dbg_leave; EXIT; //refresh
    end
  else begin
         opage := page;
         opos  := idx2;
         MenuLib2_mn_environment.do_refresh := FALSE;
       end;

  For idx := page to mnu_len2+page-1 do
    If (idx = idx2+page-1) then
      ShowCStr_clone(MenuLib2_mn_environment.v_dest,mnu_x+1,mnu_y+idx2,
                     ExpStrR(pstr(idx2+page-1)+pdes(idx2+page-1),
                     max+(Length(pstr(idx2+page-1))+Length(pdes(idx2+page-1))-
                     CStrLen(pstr(idx2+page-1)+pdes(idx2+page-1))),' '),
                     MenuLib2_mn_setting.text2_attr,MenuLib2_mn_setting.short2_attr,
                     MenuLib2_mn_setting.text_attr,MenuLib2_mn_setting.short_attr)
    else
      If mbuf[idx].use then
        ShowCStr(MenuLib2_mn_environment.v_dest,mnu_x+1,mnu_y+idx-page+1,
                 ExpStrR(pstr(idx)+pdes(idx),
                 max+(Length(pstr(idx))+Length(pdes(idx2+page-1))-
                 CStrLen(pstr(idx)+pdes(idx))),' '),
                 MenuLib2_mn_setting.text_attr,MenuLib2_mn_setting.short_attr)
      else
        ShowCStr(MenuLib2_mn_environment.v_dest,mnu_x+1,mnu_y+idx-page+1,
                 ExpStrR(pstr(idx)+pdes(idx),
                 max+(Length(pstr(idx))+Length(pdes(idx2+page-1))-
                 CStrLen(pstr(idx)+pdes(idx))),' '),
                 MenuLib2_mn_setting.disbld_attr,MenuLib2_mn_setting.disbld_attr);

  vscrollbar_pos :=
    VScrollBar(MenuLib2_mn_environment.v_dest,mnu_x+max+1,mnu_y+1,temp2,mnu_count,idx2+page-1,
               vscrollbar_pos,MenuLib2_mn_setting.menu_attr,MenuLib2_mn_setting.menu_attr);

  _dbg_leave; //EXIT //refresh
end;

File: dialogio/Menu2_refresh.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2025 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

static void Menu2_refresh (void)
{
  DBG_ENTER ("Menu2_refresh");

  if ((_mnu2.page != _mnu2.opage) || (_mnu2.idx2 != _mnu2.opos) || mn2_environment.do_refresh)
  {
    _mnu2.opage = _mnu2.page;
    _mnu2.opos  = _mnu2.idx2;
    mn2_environment.do_refresh = false;

    for (_mnu2.idx = _mnu2.page; _mnu2.idx <= _mnu2.len2 + _mnu2.page - 1; _mnu2.idx++)
    {
      if (_mnu2.idx == _mnu2.idx2 + _mnu2.page - 1)
      {
        String_t s, t;

        s = Menu2_pstr (_mnu2.idx2 + _mnu2.page - 1);
        t = Menu2_pdes (_mnu2.idx2 + _mnu2.page - 1);
        AppendString ((String *)&s, (String *)&t, 255);
        s = ExpStrR ((String *)&s, _mnu2.max + s.len - CStrLen ((String *)&s), ' ');

        Menu2_ShowCStr (mn2_environment.v_dest,
                        _mnu2.x + 1,
                        _mnu2.y + _mnu2.idx2,
                        (String *)&s,
                        mn2_setting.text2_attr,
                        mn2_setting.short2_attr,
                        mn2_setting.text_attr,
                        mn2_setting.short_attr);
      }
      else
      {
        String_t s, t;
        size_t sl;
        uint8_t attrs[2];

        s = Menu2_pstr (_mnu2.idx);
        sl = s.len;
        t = Menu2_pdes (_mnu2.idx);
        AppendString ((String *)&s, (String *)&t, 255);
        t = Menu2_pdes (_mnu2.idx2 + _mnu2.page - 1);
        s = ExpStrR ((String *)&s, _mnu2.max + sl + t.len - CStrLen ((String *)&s), ' ');

        if (_mnu2.mbuf[_mnu2.idx - 1].use)
        {
          attrs[0] = mn2_setting.text_attr;
          attrs[1] = mn2_setting.short_attr;
        }
        else
        {
          attrs[0] = mn2_setting.disbld_attr;
          attrs[1] = mn2_setting.disbld_attr;
        }

        ShowCStr (mn2_environment.v_dest,
                  _mnu2.x + 1,
                  _mnu2.y + _mnu2.idx - _mnu2.page + 1,
                  (String *)&s,
                  attrs[0],
                  attrs[1]);
      }
    }

    _mnu2.vscrollbar_pos =
      VScrollBar (mn2_environment.v_dest,
                  _mnu2.x + _mnu2.max + 1,
                  _mnu2.y + 1,
                  _mnu2.vscrollbar_size,
                  _mnu2.count,
                  _mnu2.idx2 + _mnu2.page - 1,
                  _mnu2.vscrollbar_pos,
                  mn2_setting.menu_attr,
                  mn2_setting.menu_attr);
  }

  DBG_LEAVE (); //EXIT //Menu2_refresh
}
